network/local.go

```diff
@@ -0,0 +1,459 @@
+	// and delete the remote one + close it
+	remote, ok := ccc.queues[conn.remote]
+	if !ok {
+		return
+	}
+	delete(ccc.queues, conn.remote)
+	remote.Close()
+}
+
+// Len returns how many local connections is there
+func (ccc *LocalContext) Len() int {
+	ccc.Lock()
+	defer ccc.Unlock()
+	return len(ccc.queues)
+}
+
+// LocalConn is a connection that send and receive messages through channels
+type LocalConn struct {
+	local  endpoint
+	remote endpoint
+
+	// connQueue is the part that is accesible from the LocalContext (i.e. is
+	// shared). Reason why we can't directly share LocalConn is because go test
+	// -race detects as data race (while it's *protected*)
+	*connQueue
+
+	ctx *LocalContext
+}
+
+// newLocalConn simply init the fields of a LocalConn but do not try to
+// connect. It should not be used as-is, most user wants to call NewLocalConn.
+func newLocalConn(ctx *LocalContext, local, remote endpoint) *LocalConn {
+	return &LocalConn{
+		remote:    remote,
+		local:     local,
+		connQueue: newConnQueue(),
+		ctx:       ctx,
+	}
+}
+
+// NewLocalConn returns a new channel connection from local to remote
+// Mimics the behavior of NewTCPConn => tries connecting right away.
+// It uses the default local context.
+func NewLocalConn(local, remote Address) (*LocalConn, error) {
+	return NewLocalConnWithContext(defaultLocalContext, local, remote)
+}
+
+// NewLocalConnWithContext is similar to NewLocalConn but takes a specific
+// LocalContext.
+func NewLocalConnWithContext(ctx *LocalContext, local, remote Address) (*LocalConn, error) {
+	return ctx.Connect(local, remote)
+}
+
+// Send implements the Conn interface.
+func (cc LocalConn) Send(ctx context.Context, msg Body) error {
+
+	var body Body
+	var val = reflect.ValueOf(msg)
+	if val.Kind() == reflect.Ptr {
+		val = val.Elem()
+	}
+	body = val.Interface()
+
+	var typ = TypeFromData(body)
+	nm := Packet{
+		MsgType: typ,
+		Msg:     body,
+	}
+	return cc.ctx.Send(cc.remote, nm)
+}
+
+// Receive implements the Conn interface.
+func (cc *LocalConn) Receive(ctx context.Context) (Packet, error) {
+	return cc.Pop()
+}
+
+// Local implements the Conn interface
+func (cc *LocalConn) Local() Address {
+	return cc.local.addr
+}
+
+// Remote implements the Conn interface
+func (cc *LocalConn) Remote() Address {
+	return cc.remote.addr
+}
+
+// Close implements the Conn interface
+func (cc *LocalConn) Close() error {
+	cc.connQueue.Close()
+	// close the remote conn also
+	cc.ctx.Close(cc)
+	return nil
+}
+
+// Rx implements the Conn interface
+func (cc *LocalConn) Rx() uint64 {
+	return 0
+}
+
+// Tx implements the Conn interface
+func (cc *LocalConn) Tx() uint64 {
+	return 0
+}
+
+// Type implements the Conn interface
+func (cc *LocalConn) Type() ConnType {
+	return Local
+}
+
+type connQueue struct {
+	*sync.Cond
+	queue    []Packet
+	isClosed bool
+}
+
+func newConnQueue() *connQueue {
+	return &connQueue{
+		Cond: sync.NewCond(&sync.Mutex{}),
+	}
+}
+
+func (c *connQueue) Push(p Packet) {
```

Comment:
I don't care if you call this `c`or `cq` (as long as it is consistent through the whole code). Another question regarding this line, though: `Push` will be used outside of this package?